[SPARK-48781][SQL] Add Catalog APIs for loading stored procedures

aokolnychyi opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This PR contains new connector APIs for loading stored procedures per [discussed and voted](https://lists.apache.org/thread/w586jr53fxwk4pt9m94b413xyjr1v25m) SPIP tracked in [SPARK-44167](https://issues.apache.org/jira/browse/SPARK-44167). It is a subset of changes from PR #47183.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Stored procedures are routines invoked via CALL statements. They are commonly used for encapsulating non-trivial operations and may contain multiple commands, conditional logic, and side effects. These changes are needed to allow catalogs to expose custom routines to Spark. This effort aims to enhance Spark’s ability to interact with external connectors and allow users to perform more operations in plain SQL.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

Yes, this PR adds new connector APIs.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->

N/A.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->

No.

aokolnychyi commented:
cc @cloud-fan @viirya @dongjoon-hyun @huaxingao @sunchao @HyukjinKwon @gengliangwang

aokolnychyi commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/ProcedureParameter.java:
```diff
@@ -0,0 +1,84 @@
+  /**
+   * Returns the default value of this parameter, if provided.
+   */
+  Optional<Literal<?>> defaultValue();
```
I can't say I am entirely happy with this representation of default values so ideas are welcome. My assumption is that default values are expressed using foldable expressions.

cloud-fan replied:
it depends on what we want to support. Do we want to allow `current_date()`, `rand()`, etc. to be the default value?

aokolnychyi replied:
I think there are two questions to consider:
1. Do we let connectors define arbitrary expressions as default values?
2. If we support CREATE PROCEDURE in the future, do we have a way to provide the result to connectors? 

The answer to 2 is probably yes as we can have `Supplier<Literal<?>>` that internally invokes `eval` on any foldable Catalyst expression and exposes a public `Literal` to connectors.

The answer to 1 is not so clear. It seems nice to be able to express arbitrary expressions but I am not sure it will be helpful. Also, we will have to covert those public expressions to internal Catalyst expressions to evaluate (so the list of what is supported will always be limited). We already have `Literal<?>` used `ColumnDefaultValue`.

Is there a good use case where connectors may provide arbitrary expressions as default values instead of supplying a literal?

aokolnychyi replied:
Let me explore `ColumnDefaultValue` a bit more and come back.

aokolnychyi replied:
Well, we can consider exposing SQL for default values and let Spark evaluate that. This will avoid the need to map all possible V2 expressions into Catalyst, mitigating the biggest problem I had with exposing arbitrary expression. Thoughts, @cloud-fan?

cloud-fan replied:
+1, let's use SQL string here. Spark will parse it and do necessary validation.

aokolnychyi replied:
Resolving this thread as it was addressed.

aokolnychyi commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/ProcedureParameterImpl.java:
```diff
@@ -0,0 +1,90 @@
+import org.apache.spark.sql.connector.expressions.Literal;
+import org.apache.spark.sql.types.DataType;
+
+class ProcedureParameterImpl implements ProcedureParameter {
```
These are package-private to ensure folks will use public methods in `ProcedureParameter`.

aokolnychyi commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java:
```diff
@@ -0,0 +1,67 @@
+   * @param input arguments
+   * @return the result of executing this procedure with the given arguments
+   */
+  InternalRow[] call(InternalRow input);
```
See how this is being used in PR #47183.

HyukjinKwon commented:
cc @cloud-fan and @allisonwang-db is this related to something you're working on?

aokolnychyi committed:
[SPARK-48781][SQL] Add catalog APIs for loading stored procedures

aokolnychyi commented:
I've updated this PR to address what was previously discussed. A reference implementation can be found in PR #47183.

It is ready for another round, @cloud-fan @allisonport-db @HyukjinKwon.

aokolnychyi commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/procedures/UnboundProcedure.java:
```diff
@@ -0,0 +1,44 @@
+   * @return the bound procedure best matching for the input type
+   * @throws UnsupportedOperationException if the input type is not supported (optional)
+   */
+  BoundProcedure bind(StructType inputType);
```
We can consider passing an array of `ProcedureParameter`.

aokolnychyi replied:
This is not possible because we don't know the mode of the parameters upfront.

aokolnychyi commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/procedures/SQLInvocableProcedure.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * A procedure that can be invoked via SQL.
+ * <p>
+ * This is a special type of the procedure that returns at most one result set and knows
```
An alternative to this is simply allow invoking other procedures but don't return any result.

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/ProcedureCatalog.java:
```diff
@@ -0,0 +1,39 @@
+   *
+   * @param ident a procedure identifier
+   * @return the loaded unbound procedure
+   * @throws NoSuchProcedureException if the procedure does not exist
```
Do we really need a specific exception type? I think we should give the implement freedom to throw whatever error they want, if they use Spark's error framework. e.g. implementation may throw access denied error.

When Spark invokes this method, we should throw the error as it is if it's `SparkThrowable` with the error class defined. If not, Spark should throw `FAILED_LOAD_PROCUDURE` and set the original error as the cause.

cloud-fan replied:
According to my experience, enforcing a certain exception type brings more troubles to the implementation, as it has to try-catch errors from the underlying libraries/clients that already use Spark error class, to rethrow the error with the required exception type. It also stops the implementation of customizing their own "NOT FUOND" errors.

We should also relax this exception type requirement in other DS v2 connector APIs later on, to be consistent.

aokolnychyi replied:
This makes sense for me as long as we will be consistent. I'll update this PR.

aokolnychyi replied:
Updated.

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/ProcedureCatalog.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * A catalog API for working with procedures.
+ *
+ * @since 4.1.0
```
the next release is 4.0.0

aokolnychyi replied:
Updated.

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/procedures/UnboundProcedure.java:
```diff
@@ -0,0 +1,44 @@
+   * Regardless, Spark will always perform the final validation of the arguments and rearrange them
+   * as needed based on {@link BoundProcedure#parameters() parameters}.
+   *
+   * @param inputType the input type to bind
```
We should mention how named parameters are represented, e.g. a special metadata of `StructField`.

dongjoon-hyun reviewed (commented):
To @aokolnychyi , please update all `since` tags from `4.1.0` to `4.0.0`.

dongjoon-hyun reviewed (approved):
Thank you.

+1, LGTM (except the since version issue which mentioned by @cloud-fan already).

cc @viirya , @huaxingao , @szehon-ho once more (although you are already in the loop).

aokolnychyi commented:
Thanks for reviewing, @dongjoon-hyun @cloud-fan! I've addressed the feedback and updated the PR. Once this one is in, I'll update the implementation PR.

dongjoon-hyun reviewed (approved):
+1, LGTM again. Thank you, @aokolnychyi .
Merged to master.

aokolnychyi commented:
Thank you, @dongjoon-hyun!
